### ISSUE TITLE
Coordinate and _CanBeRect type aliases too restrictive

### ISSUE THREAD

**Issue by NALStudio** [pygame-community/pygame-ce#2286]
TL;DR: The minimum requirement to be Coordinate-compliant is to conform to `collection.abc.Sequence` which is way too restrictive. This applies for _CanBeRect as well.

If I want to implement my own Vector2 implementation, due to the restrictions by the Coordinate type alias in _common.pyi, I have to at least support the functionality of `collections.abc.Sequence` for my linter to not complain which I find too restrictive as it requires functions like `index()` and `count()` to be implemented which make no sense in a Vector2 class; especially since the runtime requirements for the Coordinate type are much less strict than what the Coordinate type alias requires, as pygame _seems_ to run perfectly fine even when only `__len__` and `__getitem__` are implemented.
 
There are two ways I see which could solve this problem.

### Solution 1
Create a `typing.Protocol` for the Coordinate type alias which would require everything needed during runtime (Which presumably are `__len__` and `__getitem__`, although I'm not completely sure whether this is true). This would look something like this:
```python
class Coordinate(Protocol):
    def __getitem__(self, index: int) -> Union[int, float]: ...
    def __len__(self) -> int: ...
    # This is just an example as I'm not 100 % sure what
    # the Coordinate class' implementation requires
```

### Solution 2
Drop the requirement for `typing.Sequence` to a type which is less strict, like `typing.Collection` which would require much less to be implemented to be compliant. There is just one caveat, as `typing.Collection` doesn't require `__getitem__` to be implemented, it might necessitate other changes so that the Collection's values can be extracted from its iterator. I would see it looking something like this:
```python
Coordinate = Union[Tuple[float, float], Collection[float], Vector2]

# Or since Tuple[float, float] satisfies Collection[float]
Coordinate = Union[Collection[float], Vector2]
```

I also mentioned _CanBeRect in the title since I feel that its too restrictive as well, the solutions above can also be adapted to that.

I also thought about the possibility of me implementing this myself, but I decided against it since I'm not really familiar with the codebase of pygame and I'm a newbie contributor as I haven't contributed to open source before.

**Comment by Starbuck5** [comment]
Thanks for opening an issue.

I'm not sure what the best way forward on this is, since I'm not one of the most typing-involved people. Maybe @ankith26 or @oddbookworm would have stronger opinions.

> If I want to implement my own Vector2 implementation

This might be hypothetical, but the best was to do a custom Vector2 is to subclass pygame.Vector2. That way you can add convenience methods for yourself while keeping the speed and details of the existing Vector2

> I have to at least support the functionality of collections.abc.Sequence

I believe subclassing it would support those methods automatically if you define getitem and len

> I also thought about the possibility of me implementing this myself, but I decided against it since I'm not really familiar with the codebase of pygame and I'm a newbie contributor as I haven't contributed to open source before.

This would be a great way to start. It seems like you've already found the stubs, which are in buildconfig/stubs/pygame

**Comment by ankith26** [comment]
As I was pinged here, I wanna add that I support this proposal as you correctly highlight the issue with the existing stuff, and if you interested in implementing this yourself, I would love to take a look and guide you through it if you need help.

**Comment by ankith26** [comment]
Actually, I'd like to tackle this one myself as it is a bit more involved, I just noticed the need for related typing fixes in a couple of other places

**Comment by NALStudio** [comment]
Thanks for your answers! Ankith's PR looks to fix all the gripes I had with pygame's type stubs.

Regarding subclassing pygame's Vector2 implementation, I just felt the need to write my own implementation from scratch since I felt that the core behaviour of pygame's Vector2 was different than what I was personally looking for in a Vector2 class. Regarding performance, I wrote my Vector2 in Cython, so the performance should be comparable-ish to pygame's C implementation...
